src/bundle.rs

```diff
@@ -290,7 +290,8 @@ impl<T: Authorization, V> Bundle<T, V> {
             .filter_map(|(idx, action)| {
                 let domain = OrchardDomain::for_action(action);
                 keys.iter().find_map(move |ivk| {
-                    try_note_decryption(&domain, ivk, action)
+                    let prepared_ivk = PreparedIncomingViewingKey::new(ivk);
```

Comment:
Loop over the keys to prepare them first, so that there is only one preparation per key, not (actions × keys) preparations. (I know this is not an ideal API, but might as well avoid unnecessary inefficiency within that constraint, since it's easy to do.)